### PR TITLE
drivers: uart_gecko: added support for hardware flow control

### PR DESCRIPTION
**Pull request by tidamm**
Added support for hardware flow control in EFM32 UART driver.

Tested with samples/hello_world on efm32pg_stk3402a and further low level tests with uart_poll_in() and uart_poll_out() as in tests/drivers/uart/uart_basic_api/.

Signed-off-by: Timo Dammes <timo.dammes@lemonbeat.com>

### CODE REVIEW

**Comment by tidamm** [comment]
@galak Hello, can you provide some feedback what is missing here and prevents this from being merged? BR

File: drivers/serial/uart_gecko.c
Comment (mnkp) [inline drivers/serial/uart_gecko.c:470]: I believe the driver instantiation code would be simpler if we defined two macros `UART_RX_TX_PINS` and `UART_CTS_RTS_PINS` and then included the latter only if HW flow control is enabled for the instance. It should also make it easier to update the driver in the future to use macro based driver instantiation, similar to how other drivers, e.g. gpio_gecko.c are doing. Though in case of the uart driver this is not going to be straightforward.

File: drivers/serial/uart_gecko.c
Comment (mnkp) [inline drivers/serial/uart_gecko.c:343]: It would cleaner to write it as
```
	if (config->hw_flowcontrol == usartHwFlowControlCtsAndRts) {
```

**Review by mnkp — CHANGES_REQUESTED** [review]

**Comment by chrta** [comment]
I am taking over with this pr from my colleague. This change is based on #25977 (which is just one commit). So i can either remove #25977 or we wait with merging this until #25977 is merged.

All comments should be addressed.

File: drivers/serial/uart_gecko.c
Comment (mnkp) [inline drivers/serial/uart_gecko.c:532]: I believe we can't do this. The structure will be initialized to all 0s. The side effect of running `soc_gpio_configure` with such data will be to disable pin PA0. The code was probably supposed not to configure RTS pin at all but that's not the case. It may be easier to force user to provide location property for RTS, CTS pins. Otherwise implementation becomes complicated.

File: drivers/serial/uart_gecko.c
Comment (mnkp) [inline drivers/serial/uart_gecko.c:77]: I think the code would look cleaner if we had `bool` here. Then only on the line `usartInit.hwFlowControl = ` where we write to the usart init struct we should use the proper HAL type. The `MAP_HW_FLOWCONTROL_INIT_VAL` macro could be removed.

**Review by mnkp — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_gecko.c
Comment (chrta) [reply drivers/serial/uart_gecko.c]: Removed the redundant doc, also some doc that i think is not necessary.

File: drivers/serial/uart_gecko.c
Comment (chrta) [reply drivers/serial/uart_gecko.c]: done

File: drivers/serial/uart_gecko.c
Comment (chrta) [reply drivers/serial/uart_gecko.c]: done

File: drivers/serial/uart_gecko.c
Comment (chrta) [reply drivers/serial/uart_gecko.c]: This check is now removed. I added a check at the top of the file:
```
#if defined(UART_GECKO_HW_FLOW_CONTROL) && \
	(!defined(CONFIG_SOC_GECKO_HAS_INDIVIDUAL_PIN_LOCATION))
#error "Driver not supporting hardware flow control for this SOC"
#endif
```

File: drivers/serial/uart_gecko.c
Comment (chrta) [reply drivers/serial/uart_gecko.c]: I had to leave this COND_CODE_1 macros to be able to compile the driver  when one post has enabled hw flow control, and one other port is used without hw flow control.

The rts and cts pins are only configured, if hw flow control is enabled:
```
if (config->hw_flowcontrol) {
		soc_gpio_configure(&config->pin_rts);
		soc_gpio_configure(&config->pin_cts);
```

File: drivers/serial/uart_gecko.c
Comment (chrta) [reply drivers/serial/uart_gecko.c]: done

File: drivers/serial/uart_gecko.c
Comment (chrta) [inline drivers/serial/uart_gecko.c:468]: This ensures that location-rts and location-cts are set in the dts, when hw-flow-control is enabled.
But there are a lot of errors displayed in that case. You have to scroll up to the first error to the the message "DTS location-rts and location-cts are mandatory".
Maybe there is a better way to do this with a cleaner error message.

File: drivers/serial/uart_gecko.c
Comment (mnkp) [inline drivers/serial/uart_gecko.c:485]: `{}` is a GNU extension. Since we're planning to support other toolchains it would be better to change it to the standard `{0}`.

**Review by mnkp — COMMENTED** [review]
I overlooked one minor issue in my last review.

File: drivers/serial/uart_gecko.c
Comment (mnkp) [reply drivers/serial/uart_gecko.c]: My bad, I overlooked this. Thanks for the comment.

File: drivers/serial/uart_gecko.c
Comment (chrta) [reply drivers/serial/uart_gecko.c]: done

**Review by mnkp — APPROVED** [review]

**Comment by chrta** [comment]
@ioannisg I prefer not to review/approve code i have partly written myself. @mnkp already approved. So one review/approval from another person is missing.

**Comment by ioannisg** [comment]
> @ioannisg I prefer not to review/approve code i have partly written myself. @mnkp already approved. So one review/approval from another person is missing.

Right, Just realized that, thanks.

**Comment by ioannisg** [comment]
@mnkp assigning this PR to you.
You might have an idea of a second reviewer here.

**Comment by carlescufi** [comment]
@Oanerer and @l-alfred would you mind reviewing?

**Comment by chrta** [comment]
@mnkp I believe your approval was removed by accident, can you please re-approve?

**Review by LeGrosFaisan — APPROVED** [review]
Everything seems fine to me. I will probably add these changes to the Gecko Series 2 PR #27042 once finalised so that it also supports HW flow control.

**Review by mnkp — APPROVED** [review]

**Comment by mnkp** [comment]
> @mnkp assigning this PR to you.
> You might have an idea of a second reviewer here.

@galak could have a quick look and verify usage of DTS access macros. Otherwise, this PR LGTM.

**Review by pkaranjekar — APPROVED** [review]

File: drivers/serial/uart_gecko.c
Comment (MaureenHelm) [inline drivers/serial/uart_gecko.c:42]: @galak is there a macro we can use to detect if this property exists on any instance? I worry that this will break if there is ever a 6th instance

**Review by MaureenHelm — COMMENTED** [review]
@galak please have a look at the dts macro usages

File: drivers/serial/uart_gecko.c
Comment (galak) [reply drivers/serial/uart_gecko.c:42]: We dont have a macro right now.  Should see if this can be written using `DT_INST_FOREACH_STATUS_OKAY()`

@mbolivar-nordic any thoughts here?

File: drivers/serial/uart_gecko.c
Comment (galak) [reply drivers/serial/uart_gecko.c:468]: There isn't a better solution available today.  We'd need to look at adding some additional syntax to the binding to convey if prop X than require prop Y.

Comment (chrta) [reply drivers/serial/uart_gecko.c:468]: Converted this into an issue #27483

File: drivers/serial/uart_gecko.c
Comment (mbolivar-nordic) [reply drivers/serial/uart_gecko.c:42]: > @mbolivar-nordic any thoughts here?

@galak @MaureenHelm  yep, you can use `DT_INST_FOREACH_STATUS_OKAY()` for this as long as you're actually asking the question about status okay instances. The test as written is about any instance at all, not just status okay instances, so it's a bit different. (That said, I suspect the code as written might be wrong since I'm guessing disabled instances should not be taken into account, but that's just a hunch.)

The rest of this comment applies to the "status okay" case, but it would be trivial to add `DT_INST_FOREACH()` to devicetree.h if you really need the result for *all* instances, including disabled ones.

Any question "is `MY_TEST(inst)` true for any status okay `DT_DRV_COMPAT` instance number `inst`?" can be written as something like a [fold](https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29) using DT_INST_FOREACH_STATUS_OKAY.

A code template is:

```
#define MY_FOLD_OP(inst) MY_TEST(inst) ||
#define MY_FOLD_RESULT   DT_INST_FOREACH_STATUS_OKAY(MY_FOLD_OP) 0
```

When expanded:

```
MY_FOLD_RESULT ->
MY_FOLD_OP(0) MY_FOLD_OP(1) ... MY_FOLD_OP(n-1) 0 ->
MY_TEST(0) || MY_TEST(1) || ... MY_TEST(n-1) || 0
```

So as long as `MY_TEST` expands to an expression the preprocessor can evaluate, the entire `MY_FOLD_RESULT` does too.

In this case:

```
#define HAS_HFC_OR(inst) DT_INST_NODE_HAS_PROP(inst, hw_flow_control) ||
#define ANY_INST_HAS_HFC DT_INST_FOREACH_STATUS_OKAY(HAS_HFC_OR) 0

#if ANY_INST_HAS_HFC

...

#endif
```

@galak I will leave it up to you to decide if we should add something like `DT_ANY_INST_STATUS_OKAY(MY_TEST)` to devicetree.h based on the above.

Comment (chrta) [reply drivers/serial/uart_gecko.c]: Thank you, that is a good improvement. I included your suggestion.

File: drivers/serial/uart_gecko.c
Comment (mbolivar-nordic) [inline drivers/serial/uart_gecko.c:27]: I missed that this was a boolean property when I first responded. Boolean properties should be tested with DT_INST_PROP() or DT_PROP(), not ...NODE_HAS_PROP(). The bindings type system knows that a boolean property is true if present and false if absent.
https://docs.zephyrproject.org/latest/guides/dts/api-usage.html#property-access

Comment (chrta) [reply drivers/serial/uart_gecko.c]: done

**Review by mbolivar-nordic — COMMENTED** [review]
DT macros appear to have been addressed as discussed.

**Comment by MaureenHelm** [comment]
@chrta please have a look at the buildkite failures

**Comment by chrta** [comment]
@MaureenHelm I just did. There is still an issue:
```
$ cd /tests/kernel/common
$ west build -b efm32hg_slstk3400a --pristine

....

SRAM:        8312 B         8 KB    101.46%
/opt/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: region `SRAM' overflowed by 120 bytes
collect2: error: ld returned 1 exit status
```

This also fails for me on the current master. This SoC only has 8kByte RAM, maybe i have to reduce some thread stack sizes. I remember that was necessary in another test, also for some STM32 with only 8kB RAM.

**Comment by MaureenHelm** [comment]
> This also fails for me on the current master. This SoC only has 8kByte RAM, maybe i have to reduce some thread stack sizes. I remember that was necessary in another test, also for some STM32 with only 8kB RAM.

Try adding `min_ram` to `tests/kernel/common/testcase.yaml`

**Comment by chrta** [comment]
I opened https://github.com/zephyrproject-rtos/zephyr/issues/27583 since sanitycheck did not fail when i execute:
```
$ sanitycheck -p efm32hg_slstk3400a
```
But the build logs indicate the same error i pasted above.

**Comment by chrta** [comment]
The SRAM overflow is also present for other SoCs and in other tests and intentionally ignored by sanitycheck, so no change here necessary regarding min_ram in the testcase.yaml.

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
